src/x3f_io.c

```diff
@@ -293,15 +299,24 @@ static x3f_huffman_t *new_huffman(x3f_huffman_t **HUFP)
 
   GET4(H->version);
   GETN(H->unique_identifier, SIZE_UNIQUE_IDENTIFIER);
-  GET4(H->mark_bits);
-  GET4(H->columns);
-  GET4(H->rows);
-  GET4(H->rotation);
-  if (H->version > X3F_VERSION_2_0) {
-    GETN(H->white_balance, SIZE_WHITE_BALANCE);
-    GETN(H->extended_types, NUM_EXT_DATA);
-    for (i=0; i<NUM_EXT_DATA; i++)
-      GET4(H->extended_data[i]);
+  /* TODO: the meaning of the rest of the header for version >= 4.0
+           (Quattro) is unknown */
+  if (H->version < X3F_VERSION_4_0) {
+    GET4(H->mark_bits);
+    GET4(H->columns);
+    GET4(H->rows);
+    GET4(H->rotation);
+    if (H->version >= X3F_VERSION_2_1) {
+      int num_ext_data =
+	H->version >= X3F_VERSION_3_0 ? NUM_EXT_DATA_3_0 : NUM_EXT_DATA_2_1;
+
+      GETN(H->white_balance, SIZE_WHITE_BALANCE);
+      if (H->version >= X3F_VERSION_2_3)
+	GETN(H->color_mode, SIZE_COLOR_MODE);
+      GETN(H->extended_types, num_ext_data);
+      for (i=0; i<num_ext_data; i++)
+	GET4F(H->extended_data[i]);
+    }
```

Comment:
Versions 4.x (Quattro) have a totally different header apart from the
two first fields, identifier and version. We have so far not bothered
about trying to decipher it.

I wouldn't report it as an error since the information present in the
header is not required for functionality. Maybe it could be reported as
a warning, but I'm not sure. It's more of a todo.

On mar, 2015-07-28 at 08:14 -0700, mmroden wrote:
> In src/x3f_io.c:
> 
> > +  if (H->version < X3F_VERSION_4_0) {
> > +    GET4(H->mark_bits);
> > +    GET4(H->columns);
> > +    GET4(H->rows);
> > +    GET4(H->rotation);
> > +    if (H->version >= X3F_VERSION_2_1) {
> > +      int num_ext_data =
> > +	H->version >= X3F_VERSION_3_0 ? NUM_EXT_DATA_3_0 : NUM_EXT_DATA_2_1;
> > +
> > +      GETN(H->white_balance, SIZE_WHITE_BALANCE);
> > +      if (H->version >= X3F_VERSION_2_3)
> > +	GETN(H->color_mode, SIZE_COLOR_MODE);
> > +      GETN(H->extended_types, num_ext_data);
> > +      for (i=0; i<num_ext_data; i++)
> > +	GET4F(H->extended_data[i]);
> > +    }
> 
> Should we report an error for versions > 4.0? Or make a best-guess?
> 
> —
> Reply to this email directly or view it on GitHub.
> 
>